### ISSUE TITLE
Event Handler / Saga / Where to Write Documentation

### ISSUE THREAD

**Issue by jsonberry** [nestjs/cqrs#64]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

My main concern is understanding a good place where to do `writes`.

I'm enjoying setting up the CQRS module in my current project, but I ended up dropping event handlers because I couldn't justify a real good use for them.

From `command` handlers I am publishing events with the `eventBus`, then creating `sagas` for events and additional command handlers for those because I want to:

- Keep complicated async business logic at the `saga` level
- Consolidate writes to a unified location (command handlers)

I chose this approach so that I can onboard others into the project and if they have questions about where to do writes/reads/more complex async logic, I can direct them as such:

- writes -> command handlers
- reads -> query handlers
- more complex async logic / event sourcing-> saga

Am I missing something real obvious about event handlers? I fully understand I can inject into them and I can handle write logic there, but is there something better about the pattern that I am not groking?

Thanks!

# Edit
I realize after reading/watching more about CQRS that this is probably just me needing to understand the pattern better. 

For most of my use cases, especially for single transactions that don't require communication with other aggregates, I see that an `Event Handler` is a good place to do work. Transitioning to a `Saga` seems to come into play when there are multiple transactions that need to occur, but even then at the end of the `Saga` I'm not sure what I would need to do in the `Command Handler` at the end of a `Saga` if I'm executing transactions inside of the `Saga`.

**Comment by jsonberry** [comment]
As I've continued on in my ES/CQRS learning I've become more familiar with the patterns so where to do a write is not so elusive now.

It looks like people want to expand the official NestJS CQRS example, so maybe I can make some suggestions on that project while I'm continuing to learn.

Cheers 🍻

**Comment by subovietnam** [comment]
how using to rollback transaction in cqrs and event sourcing, in sagas when have event failed or message broker failed. Thanks all

**Comment by kodeine** [comment]
@jsonberry i'm in the same boat as you were a while ago, could you please confirm if i'm following the right pattern.

Account has many users.

User App: CreateUserCommand -> write to db, result goes to aggregate -> commit() -> UserCreatedEvent

Account App: Saga ofType UserCreatedEvent -> use the userId from event send new command -> CreateAccountCommand

i'm not using UserCreatedEventHandler here... should i be doing writes in there? so what needs to happen in command handler then...

**Comment by jsonberry** [comment]
Your pattern sounds logical, for this case either an event handler or a
Saga could work...

I would go with a Saga, because you know the next step in the flow is a
command (create account), and Sagas always end in a command being issued.
Sagas are my goto as a way to orchestrate events with commands. Their most
powerful feature IMHO is the async behavior via RxJS, when I needed to fire
off commands and then wait for success/failure events inside the same saga
to continue processing...(that’s how I handle rollback type features).

I have ended up using Event Handlers often, and I usually designated them
to encapsulate logic where a write to an aggregate or the storage of event
wasn’t going to happen... basically like side effects... logging, email/sms
notifications, sending events to other systems event pipelines like an AWS
SQS queue. They’re great for that stuff.
